src/utils/sshUtils.ts

```diff
@@ -3,32 +3,59 @@
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
 
+import { ComputeManagementClient, ComputeManagementModels } from '@azure/arm-compute';
 import * as fse from 'fs-extra';
 import * as os from "os";
 import { join } from 'path';
-import { callWithMaskHandling } from 'vscode-azureextensionui';
+import { callWithMaskHandling, IParsedError, parseError } from 'vscode-azureextensionui';
+import { IVirtualMachineWizardContext } from '../commands/createVirtualMachine/IVirtualMachineWizardContext';
 import { ext } from '../extensionVariables';
 import { localize } from '../localize';
 import { VirtualMachineTreeItem } from '../tree/VirtualMachineTreeItem';
+import { createComputeClient } from './azureClients';
 import { cpUtils } from "./cpUtils";
+import { nonNullValueAndProp } from './nonNull';
 
 export const sshFsPath: string = join(os.homedir(), '.ssh');
 
-export async function getSshKey(vmName: string, passphrase: string): Promise<string> {
+export async function getSshKey(context: IVirtualMachineWizardContext, vmName: string, passphrase: string): Promise<string> {
     return await callWithMaskHandling(
         async () => {
             const sshKeyName: string = `azure_${vmName}_rsa`;
             const sshKeyPath: string = join(sshFsPath, sshKeyName);
+            const generatedKey: string = localize('generatedKey', 'Generated public/private rsa key pair in "{0}".', sshKeyPath);
 
-            if (!await fse.pathExists(`${sshKeyPath}.pub`)) {
-                ext.outputChannel.appendLog(localize('generatingKey', 'Generating public/private rsa key pair in "{0}"...', sshKeyPath));
-                // create the .ssh folder if it doesn't exist
-                await fse.ensureDir(sshFsPath);
-                await cpUtils.executeCommand(undefined, undefined, 'ssh-keygen', '-t', 'rsa', '-b', '4096', '-f', cpUtils.wrapArgInQuotes(sshKeyPath), '-N', cpUtils.wrapArgInQuotes(passphrase));
-                ext.outputChannel.appendLog(localize('generatedKey', 'Generated public/private rsa key pair in "{0}".', sshKeyPath));
-            }
+            // create the .ssh folder if it doesn't exist
+            await fse.ensureDir(sshFsPath);
+            try {
+
+                if (!await fse.pathExists(`${sshKeyPath}.pub`)) {
+                    ext.outputChannel.appendLog(localize('generatingKey', 'Generating public/private rsa key pair in "{0}"...', sshKeyPath));
+                    await cpUtils.executeCommand(undefined, undefined, 'ssh-keygen', '-t', 'rsa', '-b', '4096', '-f', cpUtils.wrapArgInQuotes(sshKeyPath), '-N', cpUtils.wrapArgInQuotes(passphrase));
+                }
+
+                return (await fse.readFile(`${sshKeyPath}.pub`)).toString();
+
+            } catch (error) {
+                const parsedError: IParsedError = parseError(error);
```

Comment:
Any way we can be more specific about when we fallback to azure ssh generation? I figured it would _only_ be used if ssh-keygen is not on their machine. Typically, I would just run `func --version` or something to see if a cli is installed.

Reply:
Part of the reason that I want to fallback no matter what the failure is because the SSH keys are generated pretty much last (right before the VM is created).  So if it fails here, there are a several resources that are now pointless.  We could keep track of everything created and delete upon failure, I guess.

Also, as far as I can tell, `ssh-keygen` doesn't have an equivalent `func --version`.  If I try to run the command with invalid parameters, it will open the help usage document: 
![image](https://user-images.githubusercontent.com/5290572/102424811-d5e8b300-3fc0-11eb-9d34-9e22a74e536a.png)

Whereas it'd throw an error if it didn't exist.

Reply:
Actually though, we might be able just create it without a public SSH key and instruct the user to do so after creation.

Reply:
> Also, as far as I can tell, ssh-keygen doesn't have an equivalent func --version. If I try to run the command with invalid parameters, it will open the help usage document:

Yeah you might have to be more creative about detecting if ssh-keygen is installed.

> Actually though, we might be able just create it without a public SSH key and instruct the user to do so after creation.

This sounds like a good idea.

One reason I want us to be more specific about when we fall back to azure ssh keys is for predictability. If I create VMs on a regular basis through this extension, I want them to be created the same way each time. I don't want it to use an azure key if ssh-keygen fails for some random transient error

Reply:
> > Actually though, we might be able just create it without a public SSH key and instruct the user to do so after creation.
>
>This sounds like a good idea.
>
> One reason I want us to be more specific about when we fall back to azure ssh keys is for predictability. If I create VMs on a regular basis through this extension, I want them to be created the same way each time. I don't want it to use an azure key if ssh-keygen fails for some random transient error

Unfortunately, this won't work out since Azure requires you to have authentication when creating a VM. 

![image](https://user-images.githubusercontent.com/5290572/102545180-c2922200-406a-11eb-8d43-1eb6ae7b2971.png)

If you pass in an empty SSH key, it'll fail with invalid keydata.  I think an alternative is to still fallback to Azure either way, but be more explicit that either ssh-keygen wasn't found so we're falling back or ssh-keygen failed so we're falling back?

Reply:
I think it's okay to fallback to Azure if ssh-keygen isn't installed because we've seen users (okay maybe just Matt) run into that issue and it seems like a valid new-user scenario. If you want to convince me that we should fall back for _all_ errors, I would prefer to hear about other example errors you've seen in telemetry or user issues. I just don't think other errors are common and I'd rather not make an edge-case scenario _more_ complex by falling back to Azure (which the user may not want)

Reply:
Fair enough.  Seems low enough stakes for me anyway since I haven't really seen many errors in telemetry indicating the create failed due to ssh-keygen